(AKA-6) Карта с получением геолокации пользователя

K-Khud opened the pull request.

K-Khud committed:
(AKA-6) Добавила mapView

K-Khud committed:
(AKA-7) Добавила определение currentLocation.

K-Khud committed:
(AKA-7) Добавила запрос на доступ к геоданным пользователя

K-Khud committed:
(AKA-6) Поменяла текст запроса на геолокацию.

belannik commented on Map/Scenes/Map scene/MapVCExtension.swift:
```diff
@@ -0,0 +1,33 @@
+		}
+	}
+
+	private func zoomToLatestLocation(with coordinate: CLLocationCoordinate2D) {
```
Это же приватная функция класса, а не протокола?

K-Khud replied:
Исправлено.

FlightyNEO replied:
Думаю лучше назвать goToLatestLocation(with:) или showLatestLocation(with:)
zoom подразумевает увеличение/уменьшение.

K-Khud committed:
Merge branch 'develop' of https://github.com/FocusStartSpb/Map into feature/AKA-6-add-map-and-tabBar

FlightyNEO commented:
Почему у тебя названия комитов AKA-7. Это ж другая задача совсем?

FlightyNEO commented on .gitignore:
```diff
@@ -67,3 +67,5 @@ fastlane/Preview.html
 fastlane/test_output
 .DS_Store
+Map.xcodeproj/project.xcworkspace/xcuserdata/ekaterinakhudzhamkulova.xcuserdatad/UserInterfaceState.xcuserstate
+*.xcuserstate
```
Это что такое? 🤔

K-Khud replied:
попытки вылечить gitignore

FlightyNEO commented on Map/Scenes/Map scene/MapVCExtension.swift:
```diff
@@ -0,0 +1,33 @@
+//
+//  MapVCExtension.swift
```
MapVCExtension???

Файл должен называться MapViewController+CLLocationManagerDelegate

K-Khud replied:
исправлено

FlightyNEO commented on Map/Scenes/Map scene/MapVCExtension.swift:
```diff
@@ -0,0 +1,33 @@
+//  Created by Ekaterina Khudzhamkulova on 17.12.2019.
+//
+
+import Foundation
```
Можно не писать если есть import MapKit

K-Khud replied:
исправлено

FlightyNEO commented on Map/Scenes/Map scene/MapVCExtension.swift:
```diff
@@ -0,0 +1,33 @@
+	func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
+		guard let latestLocation = locations.first else { return }
+
+		if self.currentCoordinate == nil {
```
self можно не писать

K-Khud replied:
ok

K-Khud committed:
(AKA-6) Перенесла из файла расширения func zoom

FlightyNEO commented on Map/Scenes/Map scene/MapVCExtension.swift:
```diff
@@ -0,0 +1,33 @@
+		if self.currentCoordinate == nil {
+			zoomToLatestLocation(with: latestLocation.coordinate)
+		}
+		self.currentCoordinate = latestLocation.coordinate
```
и тут тоже

K-Khud replied:
ok

K-Khud commented:
> Почему у тебя названия комитов AKA-7. Это ж другая задача совсем?

просчиталась... теперь не знаю, как название кометов исправлять. Название ветки исправила.

FlightyNEO commented on Map/Scenes/Map scene/MapViewController.swift:
```diff
@@ -6,6 +6,7 @@
 //
 
 import UIKit
```
Тоже можно не писать import UIKit

FlightyNEO commented on Map/Scenes/Map scene/MapViewController.swift:
```diff
@@ -19,6 +20,13 @@ final class MapViewController: UIViewController
+	let locationManager = CLLocationManager()
+
+	let latitudinalMeters: Double = 5000
+	let longtitudalMeters: Double = 5000
```
все private должно быть

FlightyNEO commented on Map/Scenes/Map scene/MapViewController.swift:
```diff
@@ -19,6 +20,13 @@ final class MapViewController: UIViewController
+	let locationManager = CLLocationManager()
+
+	let latitudinalMeters: Double = 5000
+	let longtitudalMeters: Double = 5000
```
можно создать структуру 
`CoordinateMeters { latitude, longtitude: Double }`

И тип можно явно не указывать, если написать так:
`let longtitudalMeters = 5000.0`
А еще красивее так ) :
`let longtitudalMeters = 5_000.0`

FlightyNEO replied:
опечатка longitudinal

FlightyNEO replied:
и все приватное должно быть

MisnikovRoman replied:
> 5_000.0

о_О "красивее"?

FlightyNEO replied:
> о_О "красивее"?

Мне нравится, более наглядно, нули считать не надо )

K-Khud replied:
Сделала 5_000.0

FlightyNEO commented on Map/Scenes/Map scene/MapViewController.swift:
```diff
@@ -41,6 +52,31 @@ final class MapViewController: UIViewController
+		}
+		else if status == .authorizedAlways || status == .authorizedWhenInUse {
+			beginLocationUpdates(locationManager: locationManager)
+		}
```
Лучше через switch
<img width="679" alt="Снимок экрана 2019-12-18 в 00 18 16" src="https://user-images.githubusercontent.com/18489711/71035170-f7a02780-212b-11ea-922d-8b2edfe9fde1.png">

K-Khud replied:
сделала через if, но поменяла так, что во всех случаях, кроме .authorizedAlways и .authorizedWhenInUse, будет запрос авторизации (.notDetermoned,  .denied)

K-Khud replied:
в общем-то вернулась к твоему варианту со switch

FlightyNEO commented on Map/Scenes/Map scene/MapViewController.swift:
```diff
@@ -41,6 +52,31 @@ final class MapViewController: UIViewController
+		}
+	}
+
+	func beginLocationUpdates(locationManager: CLLocationManager) {
```
private func

K-Khud replied:
вынесла его в протокол и переименовала

FlightyNEO commented on Map/Scenes/Map scene/MapViewController.swift:
```diff
@@ -41,6 +52,31 @@ final class MapViewController: UIViewController
+	}
+	private func setupMapConstraints() {
+		mapView.translatesAutoresizingMaskIntoConstraints = false
+		mapView.topAnchor.constraint(equalTo: view.topAnchor, constant: 0).isActive = true
```
А почему тут не view.safeAreaLayoutGuide?

K-Khud replied:
чтобы карта заполняла всю верхнюю часть и игнорировала topAnchor у safeArea.

FlightyNEO commented on Map/Scenes/Map scene/MapViewController.swift:
```diff
@@ -41,6 +52,38 @@ final class MapViewController: UIViewController
 		let request = Map.SmartTargets.Request()
 		interactor.getSmartTargets(request: request)
 	}
+	private func configureLocationServices() {
```
почему во множественном числе Services?

K-Khud replied:
done

MisnikovRoman commented:
> Почему у тебя названия комитов AKA-7. Это ж другая задача совсем?

Надо исправлять

MisnikovRoman commented on Map/Scenes/Map scene/MapVCExtension.swift:
```diff
@@ -0,0 +1,27 @@
+import Foundation
+import MapKit
+
+extension MapViewController: CLLocationManagerDelegate
```
⚠️ Работу с геолокацией лучше выносить в отдельный сервис, ну и плюс если у вас архитектура VIPER, работа с геолокацией должна происходить в `Presenter`'e или в отдельном `UseCase`'e

FlightyNEO replied:
Да точно!

K-Khud committed:
(AKA-6) Переделала сохранение и обновление location через VIP

K-Khud committed:
(AKA-6)  getCurrentCoordinate переместила в interactor

K-Khud committed:
Merge branch 'develop' into feature/AKA-6-add-map-and-tabBar

# Conflicts:
#	.gitignore
#	Map.xcodeproj/project.xcworkspace/xcuserdata/ekaterinakhudzhamkulova.xcuserdatad/UserInterfaceState.xcuserstate

FlightyNEO commented on Map/Scenes/Map scene/MapViewController.swift:
```diff
@@ -19,6 +21,13 @@ final class MapViewController: UIViewController
 	private var interactor: MapBusinessLogic
 
+	let mapView = MKMapView()
+	//var currentCoordinate: CLLocationCoordinate2D?
```
удалить закомиченную строку

K-Khud replied:
ok

FlightyNEO commented on Map/Scenes/Map scene/MapViewController.swift:
```diff
@@ -41,12 +53,62 @@ final class MapViewController: UIViewController
+	}
+
+	func zoomToLatestLocation(with coordinate: CLLocationCoordinate2D) {
+	}
```
можно удалить

K-Khud replied:
ok

MisnikovRoman commented on Map/Scenes/Map scene/MapInteractor.swift:
```diff
@@ -18,6 +20,8 @@ final class MapInteractor
 	private var presenter: MapPresentationLogic
 	private var worker: DataWorker
```
А что за абстрактный `DataWorker`?

FlightyNEO replied:
Почему абстрактный? Это worker работающий с данными на диске.

MisnikovRoman replied:
Data - универсальный тип, почти Any, Worker'ом - можно назвать любой сущность, которая что-то делает. Вот и получается максимально абстрактно и непонятно

K-Khud replied:
исправили в другой задаче

MisnikovRoman commented on Map/Scenes/Map scene/MapModels.swift:
```diff
@@ -27,4 +27,22 @@ enum Map
+
+		struct Response
+		{
+			let coordinate: CLLocationCoordinate2D
```
А почему нельзя в презентер сразу передавать координаты (`CLLocationCoordinate2D`)? Зачем оборачивать в сущность с одним параметром?

FlightyNEO replied:
Архитектура Clean Swift построена на передаче данных внутри цикла view-interactor-presenter с помощью структур. 
Вью к интерактору - запрос
Интерактор к презентеру - ответ
Презентер к вью - вьюМодел

MisnikovRoman commented on Map/Scenes/Map scene/MapModels.swift:
```diff
@@ -27,4 +27,22 @@ enum Map
+
+		struct ViewModel
+		{
+			let coordinate: CLLocationCoordinate2D
```
И зачем отдельная вью-модель, которая дублирует Response?

FlightyNEO replied:
[Clean Swift](https://clean-swift.com/)

MisnikovRoman replied:
да, я и забыл, что у вас Clean Swift

MisnikovRoman commented on Map/Scenes/Map scene/MapViewController.swift:
```diff
@@ -19,6 +21,13 @@ final class MapViewController: UIViewController
 
+	let mapView = MKMapView()
+	//var currentCoordinate: CLLocationCoordinate2D?
+	let locationManager = CLLocationManager()
```
Все еще locationManager в VC? Ему тут не место

K-Khud replied:
исправлено

K-Khud committed:
(AKA-6) Реакция на статус и изменения локации  теперь в interactor

K-Khud committed:
(AKA-6) Правки по неймингу

K-Khud committed:
(AKA-6) Правки уровня доступа

K-Khud committed:
Merge branch 'develop' of https://github.com/FocusStartSpb/Map into feature/AKA-6-add-map-and-tabBar

K-Khud committed:
(AKA-6) CLLocationManager перемещен в Interactor

K-Khud committed:
Merge branch 'develop' into feature/AKA-6-add-map-and-tabBar

K-Khud committed:
(AKA-6) BugFix: отображение currentLocation

K-Khud committed:
(AKA-6) Удалила неиспользуемые методы

K-Khud committed:
(AKA-6) Остановила постоянное центрирование карты на пользователя

FlightyNEO commented on Map/Scenes/Map scene/MapInteractor.swift:
```diff
@@ -5,24 +5,44 @@
+	private func checkAuthorizationService() {
+		let status = CLLocationManager.authorizationStatus()
+		switch status {
+		case .notDetermined: locationManager.requestAlwaysAuthorization()
+		case .authorizedAlways, .authorizedWhenInUse:
+			locationManager.startUpdatingLocation()
+			presenter.beginLocationUpdates(response: Map.UpdateStatus.Response(accessToLocationApproved: true,
+																			   userCoordinate: currentCoordinate))
+		default:
+			locationManager.requestAlwaysAuthorization()
+		}
+	}
```
Минор. В .notDetermined и в default происходит одно и то же, можно case .notDetermined убрать

FlightyNEO commented on Map/Scenes/Map scene/MapInteractor.swift:
```diff
@@ -34,11 +54,32 @@ extension MapInteractor: MapBusinessLogic
+		guard let latestCoordinate = locations.first?.coordinate else { return }
+		if currentCoordinate == nil {
+			presenter.beginLocationUpdates(response: Map.UpdateStatus.Response(accessToLocationApproved: true,
+																			   userCoordinate: latestCoordinate))
```
лишний enter

Commit:
Merge branch 'develop' into feature/AKA-6-add-map-and-tabBar

K-Khud committed:
Merge branch 'develop' of https://github.com/FocusStartSpb/Map into feature/AKA-6-add-map-and-tabBar

# Conflicts:
#	Map/Scenes/Map scene/MapInteractor.swift

K-Khud committed:
(AKA-6) Пофиксила проблемки после мержа

Interactor: методы CLLocationDelegate -> в тело класса, добавила соответствие протоколу NSObject для MapInteractor.

K-Khud committed:
Merge branch 'feature/AKA-6-add-map-and-tabBar' of https://github.com/FocusStartSpb/Map into feature/AKA-6-add-map-and-tabBar

# Conflicts:
#	Map/Scenes/Map scene/MapInteractor.swift
